feat: add guidance for engineering defaults (#47)
liatrio/openo11y.dev: docs/engineering-defaults.md

# Engineering Defaults

> Much of OpenO11y's guidance is based on or enhanced by these best practices for

Reviewer: We need to add more of the intro / definitions as to what/why Engineering Defaults.

Author: Let's iterate on this with additional PRs

> Much of Open O11y's guidance is based on or enhanced by these best practices for

Reviewer 2: The definition we've used internally and with clients has historically been:

> Engineering Defaults outline principles and practices for how Product Engineering teams should chose to operate under most circumstances. 

I'd like to combine these two for consistency.

engineering teams.

## Pair Programming

Pair programming is a software development technique in which two developers work
together at one workstation. One, the driver, writes code while the other, the
observer or navigator, reviews each line of code as it is typed in. The two
developers switch roles frequently.

This method provides extremely fast feedback to developers about their code
changes and design decisions before. In turn, this prevents wasted effort and produces
cleaner, simpler code at a faster rate.

Tooling can greatly improve the developers' experience while pair programming.
Audio calls and screen sharing enable pair programming for engineers in
different physical locations. Similarly, IDE tools such as Visual Studio Code's
[Live Share][liveshare] extension allow both parties to develop code changes
simultaneously in a single workspace.

## Small Batch Delivery

> Open O11y encourages teams to deliver in small batches. Planning small portions

Reviewer 2: ```suggestion
OpenO11y encourages teams to deliver in small batches. Delivering small portions
```

of work enables teams to deploy, demo, and request feedback more frequently.
Much like pair programming, this practice aids teams in coordinating their
efforts and safeguards against individuals investing time and energy in
developing an incorrect solution.

> ## Trunk Based Development

Reviewer: I'd like to see more of the original content in this one.  It has resonated pretty well with our customers.

Author: Let's iterate on this with additional PRs

> A source-control branching model, where developers collaborate on code in a
> single branch called 'trunk'[^trunksummary]

> Open O11y highly recommends using [Trunk Based Development][trunk] over other

Reviewer 2: ```suggestion
OpenO11y highly recommends using [Trunk Based Development][trunk] over other
```

branching models. Its focus on simplicity offers many advantages:

* The 'main' branch represents the application and is always production-ready
* Smaller, short-lived branches are dramatically safer to merge
* Developers enjoy a reduced cognitive load
* CI/CD requires less maintenance and/or is much simpler

> [Gitflow][gitflow]

Reviewer 2: I seriously wonder if we should include the gitflowsucks link here.

Author: I considered it too. It has good info. And it would be funny

Reviewer 2: @RobertKelly - any objections to us including gitflowsucks.com here?

Reviewer 3: Up to you!

## CI/CD

### Continuous Integration

> Continuous Integration (CI) is a development practice where developers
> integrate code into a shared repository frequently, preferably several times a
> day. Each integration can then be verified by an automated build and automated
> tests.[^ci]

CI is a critical part of the development lifecycle and is a natural compliment
to [Small Batch Delivery](#small-batch-delivery). This provides developers with
very frequent feedback on their code changes. When problematic merges occur,
teams are quickly notified. More importantly, a bad merge can be identified
immediately, allowing the relevant team members to quickly fix the issue.

### Continuous Delivery

> Continuous Delivery can be traced back to the first principal of the Agile

Reviewer 2: CC. @ericchapman80

Manifesto[^agilemanifesto] and naturally follows Continuous Integration. It is
the practice of sequentially deploying every viable version of an application,
ideally to production. This ensures that the software can always be released at
any time.

> ### Speed

Reviewer 2: Was this Focus on Flow before?

Author: This came from the "Fast Automated Builds" portion of the Engineering Defaults google doc. There's so much overlap with ci/cd. So I moved the basic idea here.

I think we can create a better title. Automation Performance or similar.

Reviewer 2: Why not keep it as "Fast Automated Builds?"

Author: I thought the "automated" aspect was redundant with the previous sections. No good reason for changing "fast" to "speed" though.

needed. The team partners with Product to know what is expected from the system.
SLIs and SLOs are defined, dashboards created, and alerting configured. The team
knows what to do when there is an alert.

It’s much harder to add observability and supportability ‘hooks’ after a feature
is complete and usually requires significant refactoring. Putting a minimum set
of observability and runbooks/KB in place before shipping to production
increases resilience of the system, and allows for faster feedback.

The 2022 State of DevOps report found "when a service is unreliable, users won’t
benefit from pushing code faster into that fragile context." This means that
benefits of the other Sensible Defaults can be limited (or unlocked) by how the
team is Building for Production.

## Build In Security

Teams should always be concerned with security and this should affect how they
plan work items. If a work item has security implications or concerns, the same
acceptance criteria should address them. This prevents teams from separating
security needs from other work and denies the opportunity to delay the
implementation of security requirements.

> While this list is not comprehensive, it is important that:

Reviewer 2: I'm hesitant to include the list. There's SO MUCH to security.

Author: Ya I had similar feelings. That's roughly a copy/paste from the google doc

Reviewer 2: I think let's keep the list, but change the intro sentence to something like:

```suggestion
Security has many principles and rules that should be followed.
For example, NIST has several publications, include 800-53[^nist]
that dive deep into information technology security. The following
list is a starting point of a few practices and principles a team can begin
to follow without boiling the ocean.
```

The NIST reference is: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-53r5.pdf

Author: Updated.

> # Engineering Defaults

Reviewer 4: I would like to have a intro for Engineering defaults and why we recommend these and why it is important. What outcomes are these intended to accomplish @adrielp @acramsay @ericchapman80

Much of OpenO11y's guidance is based on or enhanced by these best practices for
engineering teams.

> ## Pair Programming

Reviewer 4: A bit of structure for these will help
- Definition
- How to do X (default)
- How to measure (connecting it back to metrics)
